### ISSUE TITLE
How to learn this stuff?

### ISSUE THREAD

**Issue by Leichesters** [facebookresearch/demucs#458]
So I have a basic understanding on how AI works, but I don't know how I would make demucs any better. How do you guys at Facebook learned all this stuff? I'd like to be able to extract the instrumental of every song. I'm not afraid of learning the hard stuff too. Is it enough for that to collect - let's say 10.000 songs and their instrumental and somehow "train" (whatever this means) the AI on it?

Sorry if this seems kinda strange but I don't know how to express this better.

**Comment by akx** [comment]
This repo uses pre-trained models, so you don't need to collect files or train anything – the folks at Facebook Research have done that for you.

Just install this and run

```
demucs --two-stems=vocals file1.mp3 file2.mp3 ...
```
and have some patience to get separated instrumentals and vocals for those songs.

**Comment by Leichesters** [comment]
Hey @akx I'm very thankful for the Facebook team doing the hard work - unfortunately the provided models do not properly work on the songs I listen to. That's why I was thinking about training my own model - do you know how to do that?

**Comment by akx** [comment]
There's a document linked in the readme: https://github.com/facebookresearch/demucs/blob/main/docs/training.md

**Comment by Leichesters** [comment]
Thank you, I will take a look at this. However, I'd still like to understand what actually happens under the hood. I saw this monstrous image on the README.md (https://github.com/facebookresearch/demucs/blob/main/demucs.png) and I would love to understand what it means and what my PC actually does when training and how the AI works at all :D.

**Comment by Ime-Dekker** [comment]
Hey, @Leichesters, let me try to explain a bit how demucs works. I am not an expert. So maybe I am wrong and propably there is way more going on.. But, 
here's a simplified overview of what occurs when training the model and how the AI works:

1. **Encoder**: Demucs starts with an encoder that takes the input audio waveform and processes it through a series of convolutional layers. This process essentially converts the audio into a set of features, making it easier for the model to work with. 🎛️

2. **Bi-directional LSTM**: Next, the features pass through a bi-directional LSTM (Long Short-Term Memory) layer. This recurrent layer helps capture both past and future context, enabling the model to understand temporal patterns and dependencies in the audio. 🕰️↔️🎶

3. **Decoder**: After the LSTM, a decoder reconstructs the individual audio sources from the features. It uses transposed convolutional layers (also known as deconvolution) to transform the features back into audio waveforms. 🎧

4. **Loss function**: During training, the model compares its output with the ground truth (i.e., the original isolated audio sources) using a loss function. Demucs employs a combination of L1 loss and deep clustering loss to optimize source separation. The model's parameters are updated iteratively to minimize this loss. 🔧

Your PC, when training, performs numerous calculations and adjustments to the model's weights, aiming to separate audio sources effectively. The overall process involves encoding the input, processing it through LSTM layers, and decoding it back into individual audio components. 💻🚀

This is what I understand from it.
